ext/pcre/php_pcre.c

```diff
@@ -2273,39 +2286,41 @@ static size_t preg_replace_func_impl(zval *return_value, zval *regex, zend_fcall
  */
 static void preg_replace_common(INTERNAL_FUNCTION_PARAMETERS, int is_filter)
 {
-	zval *regex, *replace, *subject, *zcount = NULL;
+	zval *zcount = NULL;
+	zend_string *regex_str;
+	HashTable *regex_ht;
+	zend_string *replace_str;
+	HashTable *replace_ht;
+	zend_string *subject_str;
+	HashTable *subject_ht;
 	zend_long limit = -1;
 	size_t replace_count = 0;
 	zend_string	*result;
 	size_t old_replace_count;
 
 	/* Get function parameters and do error-checking. */
 	ZEND_PARSE_PARAMETERS_START(3, 5)
-		Z_PARAM_ZVAL(regex)
-		Z_PARAM_ZVAL(replace)
-		Z_PARAM_ZVAL(subject)
+		Z_PARAM_STR_OR_ARRAY_HT(regex_str, regex_ht)
+		Z_PARAM_STR_OR_ARRAY_HT(replace_str, replace_ht)
+		Z_PARAM_STR_OR_ARRAY_HT(subject_str, subject_ht)
 		Z_PARAM_OPTIONAL
 		Z_PARAM_LONG(limit)
 		Z_PARAM_ZVAL(zcount)
 	ZEND_PARSE_PARAMETERS_END();
 
-	if (Z_TYPE_P(replace) != IS_ARRAY) {
-		convert_to_string_ex(replace);
-		if (Z_TYPE_P(regex) != IS_ARRAY) {
-			convert_to_string_ex(regex);
-		}
-	} else {
-		if (Z_TYPE_P(regex) != IS_ARRAY) {
-			zend_argument_type_error(1, "must be of type array when argument #2 ($replace) is an array, %s given", zend_zval_type_name(regex));
-			RETURN_THROWS();
-		}
+	/* If a replace is an array then the regex argument needs to be also an array */
+	if (replace_ht && regex_ht == NULL) {
```

Comment:
```suggestion
	if (replace_ht && !regex_ht) {
```
or both with explicit NULL comparison